monitor/block_fetcher.py

```diff
@@ -130,30 +136,38 @@ def fetch_and_insert_new_blocks(self, max_number_of_blocks=5000):
 
         # sync forwards at most up until the forward sync target, but no more than
         # max_number_of_blocks
-        max_forward_sync_blocks = max(
-            0,
-            min(
-                max_number_of_blocks - number_of_synced_blocks,
-                self.fetch_forward_sync_target() - self.head.number,
-            ),
+        max_forward_block_number = (
+            self.fetch_forward_sync_target()
+            if max_block_number is None
+            else min(self.fetch_forward_sync_target(), max_block_number)
+        )
+        max_forward_sync_blocks = max(0, max_number_of_blocks - number_of_synced_blocks)
+
+        number_of_synced_blocks += self._sync_forwards(
+            max_number_of_blocks=max_forward_sync_blocks,
+            max_block_number=max_forward_block_number,
         )
-        if max_forward_sync_blocks > 0:
-            number_of_synced_blocks += self._sync_forwards(max_forward_sync_blocks)
 
         # sync backwards until we have synced max_number_of_blocks in total or we are fully synced
         assert 0 <= number_of_synced_blocks <= max_number_of_blocks
         max_backward_sync_blocks = max_number_of_blocks - number_of_synced_blocks
-        if max_backward_sync_blocks > 0:
-            number_of_synced_blocks += self._sync_backwards(max_backward_sync_blocks)
+
+        number_of_synced_blocks += self._sync_backwards(
+            max_number_of_blocks=max_backward_sync_blocks,
+            max_block_number=max_block_number,
+        )
 
         return number_of_synced_blocks
 
     def fetch_forward_sync_target(self):
         return max(self.w3.eth.blockNumber - self.max_reorg_depth, 0)
 
-    def _sync_forwards(self, max_number_of_blocks):
+    def _sync_forwards(
+        self, *, max_number_of_blocks: int, max_block_number: int
+    ) -> int:
         block_numbers_to_fetch = range(
-            self.head.number + 1, self.head.number + 1 + max_number_of_blocks
+            self.head.number + 1,
+            min(self.head.number + 1 + max_number_of_blocks, max_block_number + 1),
```

Comment:
```suggestion
            min(self.head.number + max_number_of_blocks, max_block_number) + 1,
```

assuming we need the `+ 1` just because `range` is exclusive

Reply:
I dont know, I would leave it like this, because for me it feels different for the two cases:

- The first is basically from `(head.number +1)` and number_of_blocks long
- The second is until `max_block_number` inclusively.

So for clarity I feel we should leave it like this.

Reply:
ok, put like this this makes sense as well.